static/src/javascripts/projects/common/modules/component.spec.js

```diff
@@ -0,0 +1,426 @@
+// @flow
+
+import fetchJSON_ from 'lib/fetch-json';
+
+import { Component } from './component';
+
+const fetchJSON: JestMockFn<*, *> = (fetchJSON_: any);
+
+const mockResponse = {
+    html: '<p>html</p>',
+    other: '<p>other</p>',
+};
+
+const createComponent = (props?: Object = {}): Component => {
+    const component = new Component();
+    const defaults = {
+        endpoint: 'whatever',
+    };
+
+    Object.assign(component, defaults, props);
+
+    return component;
+};
+
+jest.mock('lib/fetch-json', () => jest.fn(() => Promise.resolve(mockResponse)));
+
+describe('Component', () => {
+    describe('fetch()', () => {
+        test('call fetched() with an endpoint', () => {
+            const component = createComponent({
+                fetched: jest.fn(),
+            });
+
+            return component.fetch().then(() => {
+                expect(component.fetched).toHaveBeenCalledWith(mockResponse);
+            });
+        });
+
+        test('not call fetched() without an endpoint', () => {
+            const component = createComponent({
+                endpoint: false,
+                fetched: jest.fn(),
+            });
+
+            return component.fetch().then(() => {
+                expect(component.fetched).not.toHaveBeenCalled();
+            });
+        });
+
+        test('extract the content of `html` in response by default', () => {
+            const component = createComponent({
+                fetched: jest.fn(),
+            });
+
+            return component.fetch().then(() => {
+                if (!component.elem) {
+                    return Promise.reject(
+                        new Error('.elem property should exist')
+                    );
+                }
+
+                if (component.elem) {
+                    expect(component.elem.tagName).toBe('P');
+                }
+
+                if (component.elem) {
+                    expect(component.elem.innerHTML).toBe('html');
+                }
+
+                expect(component.fetched).toHaveBeenCalledWith(mockResponse);
+            });
+        });
+
+        test('properly extract data from response, if key was passed', () => {
+            const component = createComponent({
+                fetched: jest.fn(),
+            });
+
+            return component.fetch(undefined, 'other').then(() => {
```

Comment:
Maybe I'm being pernickety, but it would read more natural to me if we passed `null` rather than `undefined` here. I prefer to see `undefined` used as a comparator rather than a value to be passed explicitly. Is it just me?

Reply:
Also [this annotation](https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/common/modules/component.js#L97) made me 🤔 
```
parent: HTMLElement | ?Node
```

Reply:
It shouldn't be nullable at all, because

```js
bonzo(parent)[this.manipulationType](this.elem);
```

would break. Although my suspicion is, this gets never called anywhere and could be removed. Will check.